scheduler/src/cook/mesos/util.clj

```diff
@@ -362,51 +395,55 @@
                (filter #(not (:job/custom-executor %))))]
       (->>
         (cond->> jobs
                  (instance-states state) (filter #(= state (job-ent->state %)))
                  name-filter-fn (filter #(name-filter-fn (:job/name %))))
-        (take limit)))))
+        (take limit)
+        doall))))
 
-;; This is a separate query from completed jobs because most running/waiting jobs
-;; will be at the end of the time range, so the query is somewhat inefficent.
-;; The standard datomic query performs reasonably well on the smaller set of
-;; running and waiting jobs, so it's implemented that way to keep things simple.
 (defn get-active-jobs-by-user-and-state
   "Returns all jobs for a particular user in the specified state
-   and timeframe, without a custom executor.
-   Note that this query is not performant for completed jobs, use
-   get-completed-jobs-by-user instead."
-  [db user start end state name-filter-fn]
+   and timeframe, without a custom executor. This query looks for all
+   jobs by job state only (i.e., no 'success' or 'failed')"
+  [db user start end state limit name-filter-fn]
   (let [state-keyword (case state
                         "running" :job.state/running
                         "waiting" :job.state/waiting)
-        jobs
-        (timers/time!
-          (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
-          (->> (q '[:find [?j ...]
-                    :in $ ?user ?state ?start ?end
-                    :where
-                    [?j :job/state ?state]
-                    [?j :job/user ?user]
-                    [?j :job/submit-time ?t]
-                    [(<= ?start ?t)]
-                    [(< ?t ?end)]
-                    [?j :job/custom-executor false]]
-                  db user state-keyword start end)
-               (map (partial d/entity db))))]
-    (cond->> jobs name-filter-fn (filter #(name-filter-fn (:job/name %))))))
+        millis-diff (max 0 (- (.getTime end) (.getTime start) 60000)) ; 60000ms is so that we get queries that are just a tiny bit over a bucket.
+        hours-diff (/ millis-diff 3600000)
+        ; A histogram bucket, so that we can see the query time distribution as a function of the range of time being searched over.
+        bucket (cond
+                 (<= hours-diff 6) 6
+                 (<= hours-diff 12) 12
+                 (<= hours-diff 24) 24
+                 (<= hours-diff 36) 36
+                 (<= hours-diff 48) 48
+                 (<= hours-diff 72) 72
+                 :else 999)
+        ; If we have a name filter, we cannot push the limit into the query, but have to do it after the filter.
+        adaptedlimit (if name-filter-fn nil limit)
+        jobs (get-jobs-by-user-and-state-and-submit-by-state db user start end state-keyword adaptedlimit bucket)]
+    (cond->> jobs
+             name-filter-fn (filter #(name-filter-fn (:job/name %)))
+             name-filter-fn (take limit))))
```

Comment:
Should this be checking if there is limit as opposed to name-filter-fn?

Reply:
Oh, I see. I didn't read the comment on adaptedlimit.

I would recommend not doing the limit in the `get-jobs-by-user-and-state-and-submit-by-state` in any case and handling it out here in the caller. As it stands, you are required to take some odd steps to make it work (the adaptedlimit) which wouldn't be necessary if you pushed off the functionality until later.

Reply:
See above response.